knimin/lib/data_access.py

```diff
@@ -823,9 +823,114 @@ def pulldown(self, barcodes, blanks):
                                              for h in headers]))
             metadata[survey] = '\n'.join(survey_md).encode('utf-8')
 
-        failures = sorted(set(barcodes) - barcodes_seen)
+        failures = set(barcodes) - barcodes_seen
 
-        return metadata, failures
+        return metadata, self._explain_pulldown_failures(failures)
+
+    def _explain_pulldown_failures(self, barcodes):
+        """Builds failure reason list for barcodes passed
+
+        Parameters
+        ----------
+        barcodes : list of str
+            Barcodes to explain failure for
+
+        Returns
+        -------
+        list of tuples of str
+            list in the form [(barcode, reason), (barcode, reason), ...]
+        """
+        # if empty list passed, don't touch database
+        if len(barcodes) == 0:
+            return []
+
+        def update_failure(sql, all_bc, reason):
+            hold = [x[0] for x in
+                    self._con.execute_fetchall(sql, [tuple(all_bc)])]
+            remaining_bc = all_bc.difference(hold)
+            return [(bc, reason) for bc in hold], remaining_bc
+
+        failure_info = []
```

Comment:
if this was a `dict`, then `update_failures` is not nearly as confusing as you'll be able to just do set operations against `failure_info`. basically, there wouldn't be a need to return `remaining_bc` and calls would then look like:

```python
failure_info.update(update_failures(sql, reason))
current_barcodes = current_barcodes - failure_info
```